### DISCUSSION TITLE
Passing tags in Runner Class Dynamically

### DISCUSSION THREAD

**Discussion by Goku768collab** [cucumber/common#2219] [General]
Hey Everyone,

I am trying to execute my runner class by accepting tags from a .properties file. However I am unable to do this as cucumber thinks I haven't specified any tags and runs all my tests. I have given below the Runner class and configuration reader class which reads the values from properties.file. How can I resolve this?

Runner class:

import org.junit.runner.RunWith;
import io.cucumber.junit.Cucumber;
import io.cucumber.junit.CucumberOptions;
import //ConfigurationReader;

@RunWith(Cucumber.class)
@CucumberOptions(
        features = "src/test/resources/features",
        glue = "//my Stepdefenitions",
        dryRun = false,
      
)
public class CukesRunner {
    static {
        String tags = ConfigurationReader.getTags();
        if (tags != null && !tags.isEmpty()) {
            System.setProperty("cucumber.filter.tags", tags);
        }
    }
}

Configuration Reader:
public class ConfigurationReader {
    private static Properties properties = new Properties();

    static {
        loadProperties();
    }

    private static void loadProperties() {
        String configFilePath = System.getProperty("config.file", "configuration.properties");

        try (InputStream file = new FileInputStream(configFilePath)) {
            properties.load(file);
        } catch (IOException e) {
            System.out.println("Error occurred while reading configuration file: " + configFilePath);
            e.printStackTrace();
        }
    }

    public static String getProperty(String key) {
        return properties.getProperty(key);
    }

    public static String getTags() {
        return properties.getProperty("tags", "@test");
    }
}

**Comment by mpkorstanje** [comment]
You're currently using `cucumber-junit` which integrates Cucumber with JUnit 4. To make use of `System.setProperty` you have to ensure that it called before Cucumber reads the properties. But with JUnit 4 that is quite hard to do.

Instead if you upgrade to JUnit 5, and switch to using the [`cucumber-junit-platform-engine`](https://github.com/cucumber/cucumber-jvm/tree/main/cucumber-junit-platform-engine) in combination with the `junit-platform-suite` then you can use the [`@ConfigurationParametersResource`](https://junit.org/junit5/docs/current/api/org.junit.platform.suite.api/org/junit/platform/suite/api/ConfigurationParametersResource.html) annotation to configure the property file from which the tags should be read. 

You can find a minimal example project in the [cucumber-java-skeleton](https://github.com/cucumber/cucumber-java-skeleton).

--- 

Alternatively, you can use a Maven Profile (or the Gradle equivalent) to set `cucumber.filter.tags` before the test run starts.

**Reply by Goku768collab** [reply]
If I use a maven profile to pass the tags, wouldn't it mean people who want to run this will have to go inside the pom.xml rather than just editing a file? I am trying to minimalize user interaction with code

**Reply by mpkorstanje** [reply]
It depends. If you have predefined all the relevant profiles then users can select a profile with `mvn -Pprofile-name`.  Or if you want users to edit a file, you could use the `cucumber.properties` file. https://github.com/cucumber/cucumber-jvm/tree/main/cucumber-core#properties-environment-variables-system-options
